blssig/src/verifier/mod.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use bls_signatures::{PublicKey, Serialize, Signature, verify_messages};
+use filecoin_f3_gpbft::PubKey;
+use filecoin_f3_gpbft::api::Verifier;
+use hashlink::LruCache;
+use parking_lot::RwLock;
+use thiserror::Error;
+
+use crate::bdn::BDNAggregation;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum BLSError {
+    #[error("empty public keys provided")]
+    EmptyPublicKeys,
+    #[error("invalid public key length: expected {BLS_PUBLIC_KEY_LENGTH} bytes, got {0}")]
+    InvalidPublicKeyLength(usize),
+    #[error("failed to deserialize public key")]
+    PublicKeyDeserialization,
+    #[error("invalid signature length: expected {BLS_SIGNATURE_LENGTH} bytes, got {0}")]
+    InvalidSignatureLength(usize),
+    #[error("failed to deserialize signature")]
+    SignatureDeserialization,
+    #[error("BLS signature verification failed")]
+    SignatureVerificationFailed,
+    #[error("mismatched number of public keys and signatures: {pub_keys} != {sigs}")]
+    LengthMismatch { pub_keys: usize, sigs: usize },
+}
+
+/// BLS signature verifier using BDN aggregation scheme
+///
+/// This verifier implements the same scheme used by `go-f3/blssig`, with:
+/// - BLS12_381 curve
+/// - G1 for public keys, G2 for signatures  
+/// - BDN aggregation for rogue-key attack prevention
+/// - Point cache for performance optimization
+pub struct BLSVerifier {
+    /// Cache for deserialized public key points to avoid expensive repeated operations
+    point_cache: RwLock<LruCache<Vec<u8>, PublicKey>>,
+}
+
+impl Default for BLSVerifier {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// BLS12-381 public key length in bytes
+const BLS_PUBLIC_KEY_LENGTH: usize = 48;
+
+/// BLS12-381 signature length in bytes
+const BLS_SIGNATURE_LENGTH: usize = 96;
+
+/// Maximum number of cached public key points to prevent excessive memory usage
+const MAX_POINT_CACHE_SIZE: usize = 10_000;
+
+impl BLSVerifier {
+    pub fn new() -> Self {
+        Self {
+            point_cache: RwLock::new(LruCache::new(MAX_POINT_CACHE_SIZE)),
+        }
+    }
+
+    /// Verifies a single BLS signature
+    fn verify_single(&self, pub_key: &PubKey, msg: &[u8], sig: &[u8]) -> Result<(), BLSError> {
+        // Validate input lengths
+        if pub_key.0.len() != BLS_PUBLIC_KEY_LENGTH {
+            return Err(BLSError::InvalidPublicKeyLength(pub_key.0.len()));
+        }
+        if sig.len() != BLS_SIGNATURE_LENGTH {
+            return Err(BLSError::InvalidSignatureLength(sig.len()));
+        }
+
+        // Get cached public key
+        let pub_key = self.get_or_cache_public_key(&pub_key.0)?;
+
+        // Deserialize signature
+        let signature =
+            Signature::from_bytes(sig).map_err(|_| BLSError::SignatureDeserialization)?;
+
+        // Verify using bls-signatures
+        let messages = vec![msg];
+        let public_keys = vec![pub_key];
```

Comment:
We could avoid heap allocations here
```suggestion
        let messages = [msg];
        let public_keys = [pub_key];
```

Reply:
Fixed here: https://github.com/ChainSafe/rust-f3/pull/27/commits/6bccb280b2bc5896fe515a94dc63ceea17fbdada